File: backend/routes/qrcodes.ts
import express, { Router, Request, Response } from 'express';
import db from '../config/database';

const router: Router = express.Router();

// Get QR code by name
router.get('/:qrName', async (req: Request<{ qrName: string }>, res: Response) => {
  try {
    const { qrName } = req.params;
    
    const result = await db.query(
      `SELECT qr.*, t.name as template_name, t.description as template_description, 
              t.image_url as template_image_url, t.price as template_price
       FROM qr_codes qr
       LEFT JOIN templates t ON qr.template_id = t.id
       WHERE qr.qr_name = $1`,
      [qrName.toLowerCase()]
    );
    
    if (result.rows.length === 0) {
      return res.status(404).json({ 
        success: false, 
        error: 'QR code not found' 
      });
    }
    
    const qrCode = result.rows[0];
    
    // Parse content lines
    const contentLines = qrCode.content.split('\n').filter((line: string) => line.trim());
    
    return res.json({
      success: true,
      qrCode: {
        id: qrCode.id,
        qrName: qrCode.qr_name,
        fullUrl: qrCode.full_url,
        content: qrCode.content,
        contentLines: contentLines,
        template: {
          id: qrCode.template_id,
          name: qrCode.template_name,
          description: qrCode.template_description,
          imageUrl: qrCode.template_image_url,
          price: qrCode.template_price,
        },
        createdAt: qrCode.created_at,
      }
    });
  } catch (error) {
    const err = error as Error;
    console.error('Error fetching QR code:', err);
    return res.status(500).json({ success: false, error: err.message });
  }
});

export default router;

File: backend-golang/internal/handlers/qrcodes.go
package handlers

import (
	"context"
	"net/http"
	"strings"

	"github.com/go-chi/chi/v5"

	"inanhxink/backend-golang/internal/config"
)

// GET /api/qrcodes/:qrName
func GetQRCode(w http.ResponseWriter, r *http.Request) {
	qrName := strings.ToLower(chi.URLParam(r, "qrName"))

	rows, err := config.DB.Query(context.Background(), `
		SELECT qr.*, t.name AS template_name, t.description AS template_description,
			t.image_url AS template_image_url, t.price AS template_price
		FROM qr_codes qr
		LEFT JOIN templates t ON qr.template_id = t.id
		WHERE qr.qr_name = $1`, qrName)
	if err != nil {
		InternalError(w, err)
		return
	}
	row, err := CollectOne(rows)
	if err != nil || row == nil {
		JSON(w, 404, map[string]any{"success": false, "error": "QR code not found"})
		return
	}

	content, _ := row["content"].(string)
	var contentLines []string
	for _, line := range strings.Split(content, "\n") {
		if strings.TrimSpace(line) != "" {
			contentLines = append(contentLines, strings.TrimSpace(line))
		}
	}

	var templateImgURL any
	if v, ok := row["template_image_url"].(string); ok {
		templateImgURL = config.CdnStr(v)
	}

	OK(w, map[string]any{
		"success": true,
		"qrCode": map[string]any{
			"id":           row["id"],
			"qrName":       row["qr_name"],
			"fullUrl":      row["full_url"],
			"content":      content,
			"contentLines": contentLines,
			"template": map[string]any{
				"id":          row["template_id"],
				"name":        row["template_name"],
				"description": row["template_description"],
				"imageUrl":    templateImgURL,
				"price":       row["template_price"],
			},
			"createdAt": row["created_at"],
		},
	})
}
